src/geometry.cpp

```diff
@@ -0,0 +1,496 @@
+#include "openmc/geometry.h"
+
+#include <array>
+#include <sstream>
+
+#include "openmc/cell.h"
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/lattice.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/string_utils.h"
+#include "openmc/surface.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+
+namespace model {
+
+int root_universe {-1};
+int n_coord_levels;
+
+std::vector<int64_t> overlap_check_count;
+
+} // namespace model
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+bool check_cell_overlap(Particle* p, bool error)
+{
+  int n_coord = p->n_coord_;
+
+  // Loop through each coordinate level
+  for (int j = 0; j < n_coord; j++) {
+    Universe& univ = *model::universes[p->coord_[j].universe];
+
+    // Loop through each cell on this level
+    for (auto index_cell : univ.cells_) {
+      Cell& c = *model::cells[index_cell];
+      if (c.contains(p->coord_[j].r, p->coord_[j].u, p->surface_)) {
+        if (index_cell != p->coord_[j].cell) {
+          if (error) {
+            std::stringstream err_msg;
+            err_msg << "Overlapping cells detected: " << c.id_ << ", "
+                    << model::cells[p->coord_[j].cell]->id_ << " on universe "
+                    << univ.id_;
+            fatal_error(err_msg);
+          }
+          return true;
+        }
+        ++model::overlap_check_count[index_cell];
+      }
+    }
+  }
+
+  return false;
+}
+
+//==============================================================================
+
+bool
+find_cell_inner(Particle* p, const NeighborList* neighbor_list)
```

Comment:
This function is a bit too big with a lot of branching, can it be factored into a simpler one using helper functions? This will help reduce the cognitive load required for future readers and maintainers of your code.